keps/sig-cli/kubectl-apply-prune-enhancements.md

```diff
@@ -0,0 +1,273 @@
+- Delete objects at proper time. Assume the set of objects to be deleted is available, when to delete them is another challenge. Those objects could be still needed during rolling out other objects. One example is a ConfigMap object being referred in a Deployment object. When rolling update happens, new pods refer to new ConfigMap and old pods are still running and referring to old ConfigMap. The old ConfigMap can't be deleted until the rolling up successfully finishes.
+
+In this design, we focus on a solution for the first challenge. For the second challenge, we can blacklist certain types such as ConfigMaps and Secrets for deletion. Generally deleting objects at proper time can be better addressed by Garbage Collection.
+
+### limitation of Alpha prune
+
+The Alpha prune works by label selector:
+```
+  # Apply the configuration in manifest.yaml that matches label app=nginx and delete all the other resources that are not in the file and match label app=nginx.
+  kubectl apply --prune -f manifest.yaml -l app=nginx
+```
+Internally, `kubectl apply` takes following logic to perform pruning.
+- For one namespace, it queries all types of resources except [some hard coded ones](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/apply/apply.go#L632)
+to find objects that matches the label. The number of queries within one namespace is about the number of API resources installed in the cluster.
+- Among the matched objects, skip the ones without [last-applied-configuration annotation](https://kubernetes.io/docs/concepts/overview/object-management-kubectl/declarative-config/#how-apply-calculates-differences-and-merges-changes).
+- Among the rest objects, delete any object that is not present in the new apply set.
+- Repeat this for all visited namespaces. Overall, the number of queries is propotional to the number of API resources installed in the cluster.
+
+The limitations are mainly from three aspects:
+- Objects might be deleted unintentionally when different directories use the same labels
+- Some types are hard-coded to avoid being pruned.
+- The query number is neither constant nor linear to the number of resources being applied/pruned. 
+
+### A suggested solution
+
+#### Apply-Config Object
+To solve the first challenge, a concept of apply-config object is proposed. An apply-config object is an object that contains the GKN information for all the objects that are to be applied. The apply-config object is not generated by kubectl when applying is running. Instead, it should be generated by other tools such as kustomize, Helm. When generating resources to be consumed by `kubectl apply`, those tools are responsible to generate the associated apply-config object to utilize better pruning.
+
+An object is identified as an apply-config object when it has a special label `kubectl.kubernetes.io/apply-cfg: "true"`. The GKN information for all other objects are stored in annotations. Here is an apply-config object example.
```

Comment:
Why not make a CRD for this?

Reply:
We don't choose CRD for following reasons:
- The CRD need to be installed and not available on every cluster
- Whatever CRD we create for this, that type is going to be hard-coded in kubectl. So kubectl can recognize it.
- Some CRDs are already with the concept of a root object. By adding a label and annotation to it, pruning will work.

Reply:
@Liujingfang1 Why does this need to be a hard coded solution in kubectl? Is there supporting details, issues, or experience reports on this? I ask to understand the underlying user need.